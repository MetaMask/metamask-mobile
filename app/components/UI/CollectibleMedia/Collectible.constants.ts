import Device from '../../../util/device';

// eslint-disable-next-line import/prefer-default-export
export const MEDIA_WIDTH_MARGIN = Device.isMediumDevice() ? 32 : 0;
