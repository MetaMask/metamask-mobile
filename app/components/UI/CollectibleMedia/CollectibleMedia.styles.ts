import { StyleSheet } from 'react-native';
import scaling from '../../../util/scaling';
import Device from '../../../util/device';

import { MEDIA_WIDTH_MARGIN } from './Collectible.constants';

const createStyles = (colors: any) =>
  StyleSheet.create({
    container(backgroundColor: string) {
      return {
        flex: 0,
        borderRadius: 12,
        backgroundColor: `#${backgroundColor}`,
      };
    },
    tinyImage: {
      width: 32,
      height: 32,
    },
    smallImage: {
      width: 50,
      height: 50,
    },
    bigImage: {
      height: 260,
      width: 260,
    },
    cover: {
      height: scaling.scale(Device.getDeviceWidth() - MEDIA_WIDTH_MARGIN, {
        baseModel: 2,
      }),
    },
    image: {
      borderRadius: 12,
    },
    textContainer: {
      alignItems: 'center',
      justifyContent: 'flex-start',
      backgroundColor: colors.background.alternative,
      borderRadius: 8,
    },
    textWrapper: {
      textAlign: 'center',
      marginTop: 16,
    },
    textWrapperIcon: {
      textAlign: 'center',
      fontSize: 18,
      marginTop: 16,
    },
    mediaPlayer: {
      minHeight: 10,
    },
    imageFallBackTextContainer: StyleSheet.absoluteFillObject,
    imageFallBackShowContainer: {
      bottom: 32,
    },
    imageFallBackShowContainerCover: {
      bottom: 48,
    },
    // eslint-disable-next-line react-native/no-color-literals
    imageFallBackText: {
      textAlign: 'center',
      marginTop: 16,
      color: 'black',
    },
    imageFallBackShowText: {
      alignSelf: 'center',
    },
  });

export default createStyles;
