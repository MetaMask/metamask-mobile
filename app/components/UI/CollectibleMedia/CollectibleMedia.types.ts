import { ViewStyle } from 'react-native';

export interface Collectible {
  name: string;
  tokenId: number;
  image: string;
  imagePreview: string;
  address: string;
  animation?: string;
  backgroundColor: string;
  tokenURI: string;
  contractName?: string;
  standard: string;
  error: string | undefined;
}

export interface CollectibleMediaProps {
  collectible: Collectible;
  tiny?: boolean;
  small?: boolean;
  big?: boolean;
  renderAnimation?: boolean;
  cover?: boolean;
  style?: ViewStyle;
  onClose?: () => void;
  onPressColectible?: () => void;
}
