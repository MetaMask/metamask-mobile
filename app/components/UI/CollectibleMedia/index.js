import React, { useCallback, useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { StyleSheet, View, ViewPropTypes } from 'react-native';
import RemoteImage from '../../Base/RemoteImage';
import MediaPlayer from '../../Views/MediaPlayer';
import scaling from '../../../util/scaling';
import Text from '../../Base/Text';
import Device from '../../../util/device';
import { useTheme } from '../../../util/theme';

const MEDIA_WIDTH_MARGIN = Device.isMediumDevice() ? 32 : 0;

const createStyles = (colors) =>
  StyleSheet.create({
    container(backgroundColor) {
      return {
        flex: 0,
        borderRadius: 12,
        backgroundColor: `#${backgroundColor}`,
      };
    },
    tinyImage: {
      width: 32,
      height: 32,
    },
    smallImage: {
      width: 50,
      height: 50,
    },
    bigImage: {
      height: 260,
      width: 260,
    },
    cover: {
      height: scaling.scale(Device.getDeviceWidth() - MEDIA_WIDTH_MARGIN, {
        baseModel: 2,
      }),
    },
    image: {
      borderRadius: 12,
    },
    textContainer: {
      alignItems: 'center',
      justifyContent: 'center',
      backgroundColor: colors.background.alternative,
      borderRadius: 8,
    },
    textWrapper: {
      textAlign: 'center',
    },
    textWrapperIcon: {
      textAlign: 'center',
      fontSize: 18,
    },
    mediaPlayer: {
      minHeight: 10,
    },
  });

/**
 * View that renders an ERC-721 Token image
 */
export default function CollectibleMedia({
  collectible,
  renderAnimation,
  style,
  tiny,
  small,
  big,
  cover,
  onClose,
}) {
  const [sourceUri, setSourceUri] = useState(null);
  const { colors } = useTheme();
  const styles = createStyles(colors);

  const fallback = () => setSourceUri(null);

  useEffect(() => {
    const { image, imagePreview, address } = collectible;
    if (address) {
      if (small && imagePreview && imagePreview !== '')
        setSourceUri(imagePreview);
      else setSourceUri(image);
    }
  }, [collectible, small, big, setSourceUri]);

  const renderMedia = useCallback(() => {
    if (
      renderAnimation &&
      collectible.animation &&
      collectible.animation.includes('.mp4')
    ) {
      return (
        <MediaPlayer
          onClose={onClose}
          uri={collectible.animation}
          style={[styles.mediaPlayer, cover && styles.cover, style]}
        />
      );
    } else if (sourceUri) {
      /*
       * the tiny boolean is used to indicate when the image is the NFT source icon
       */
      return (
        <RemoteImage
          fadeIn
          resizeMode={'contain'}
          source={{ uri: sourceUri }}
          style={[
            styles.image,
            tiny && styles.tinyImage,
            small && styles.smallImage,
            big && styles.bigImage,
            cover && styles.cover,
            style,
          ]}
          onError={fallback}
        />
      );
    }
    return (
      <View
        style={[
          styles.textContainer,
          style,
          tiny && styles.tinyImage,
          small && styles.smallImage,
          big && styles.bigImage,
          cover && styles.cover,
        ]}
      >
        <Text
          big={big}
          small={tiny || small}
          style={tiny ? styles.textWrapperIcon : styles.textWrapper}
        >
          {tiny
            ? collectible.name[0] || 'C'
            : `${collectible.name || ''} #${collectible.tokenId}`}
        </Text>
      </View>
    );
  }, [
    collectible,
    sourceUri,
    onClose,
    renderAnimation,
    style,
    tiny,
    small,
    big,
    cover,
    styles,
  ]);

  return (
    <View style={styles.container(collectible.backgroundColor)}>
      {renderMedia()}
    </View>
  );
}

CollectibleMedia.propTypes = {
  /**
   * Collectible object (in this case ERC721 token)
   */
  collectible: PropTypes.object,
  /**
   * Whether is tiny size or not
   */
  tiny: PropTypes.bool,
  /**
   * Whether is small size or not
   */
  small: PropTypes.bool,
  /**
   * Whether is big size or not
   */
  big: PropTypes.bool,
  /**
   * Whether render animation or not, if any
   */
  renderAnimation: PropTypes.bool,
  /**
   * Whether the media should cover the whole screen width
   */
  cover: PropTypes.bool,
  /**
   * Custom style object
   */
  style: ViewPropTypes.style,
  /**
   * On close callback
   */
  onClose: PropTypes.func,
};
