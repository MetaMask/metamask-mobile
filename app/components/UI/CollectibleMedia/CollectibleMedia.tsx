/* eslint-disable react/prop-types */
import React, { useCallback, useEffect, useState } from 'react';
import { View } from 'react-native';
import RemoteImage from '../../Base/RemoteImage';
import MediaPlayer from '../../Views/MediaPlayer';
import Text from '../../Base/Text';
import { useTheme } from '../../../util/theme';
import { isIPFSUri } from '../../../util/general';
import { useSelector } from 'react-redux';
import {
  selectDisplayNftMedia,
  selectIsIpfsGatewayEnabled,
} from '../../../selectors/preferencesController';
import createStyles from './CollectibleMedia.styles';
import { CollectibleMediaProps } from './CollectibleMedia.types';
import NftFallbackImage from '../../../../docs/assets/nft-fallback.png';
import {
  ButtonSize,
  ButtonVariants,
} from '../../../component-library/components/Buttons/Button';
import Button from '../../../component-library/components/Buttons/Button/Button';
import { strings } from '../../../../locales/i18n';
import { useNavigation } from '@react-navigation/native';
import Routes from '../../../constants/navigation/Routes';

const CollectibleMedia: React.FC<CollectibleMediaProps> = ({
  collectible,
  renderAnimation,
  style,
  tiny,
  small,
  big,
  cover,
  onClose,
  onPressColectible,
}) => {
  const [sourceUri, setSourceUri] = useState<string | null>(null);
  const { colors } = useTheme();
  const isIpfsGatewayEnabled = useSelector(selectIsIpfsGatewayEnabled);
  const displayNftMedia = useSelector(selectDisplayNftMedia);
  const { navigate } = useNavigation();

  const styles = createStyles(colors);

  const fallback = useCallback(() => setSourceUri(null), []);

  useEffect(() => {
    const { image, imageOriginal, imagePreview, address } = collectible;
    if (address) {
      if (small && imagePreview && imagePreview !== '')
        setSourceUri(imagePreview);
      else setSourceUri((image || imageOriginal) ?? null);
    }
  }, [collectible, small, big, setSourceUri]);

  const formatTokenId = (tokenId: number) => {
    if (tokenId.toString().length > 9) {
      return tokenId.toString().replace(/^(..).*?(.{4})$/, '$1...$2');
    }
    return tokenId;
  };

  const pressNft = useCallback(() => {
    if (cover) {
      if (sourceUri && !isIPFSUri(sourceUri)) {
        navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
          screen: Routes.SHEET.SHOW_NFT_DISPLAY_MEDIA,
        });
      } else if (isIPFSUri(sourceUri) || isIPFSUri(collectible.tokenURI)) {
        navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
          screen: Routes.SHEET.SHOW_IPFS,
        });
      } else {
        navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
          screen: Routes.SHEET.SHOW_NFT_DISPLAY_MEDIA,
        });
      }
    } else {
      onPressColectible?.();
    }
  }, [cover, navigate, onPressColectible, sourceUri, collectible.tokenURI]);

  const renderFallback = useCallback(
    (isImageSourcePossiblyAvailable: boolean) =>
      isImageSourcePossiblyAvailable ? (
        <View>
          <RemoteImage
            source={NftFallbackImage}
            style={[
              styles.textContainer,
              style,
              tiny && styles.tinyImage,
              small && styles.smallImage,
              big && styles.bigImage,
              cover && styles.cover,
            ]}
            testID="fallback-nft-ipfs"
          />
          <View style={styles.imageFallBackTextContainer}>
            <Text style={styles.imageFallBackText}>
              {collectible.tokenId
                ? ` #${formatTokenId(collectible.tokenId)}`
                : ''}
            </Text>
          </View>
          {(onPressColectible || cover) && (
            <View
              style={
                cover
                  ? styles.imageFallBackShowContainerCover
                  : styles.imageFallBackShowContainer
              }
            >
              <Button
                variant={ButtonVariants.Link}
                style={styles.imageFallBackShowText}
                onPress={pressNft}
                label={strings('choose_password.show')}
                size={cover ? ButtonSize.Lg : ButtonSize.Auto}
              />
            </View>
          )}
        </View>
      ) : (
        <View
          style={[
            styles.textContainer,
            style,
            tiny && styles.tinyImage,
            small && styles.smallImage,
            big && styles.bigImage,
            cover && styles.cover,
          ]}
          testID="fallback-nft-with-token-id"
        >
          <Text
            big={big}
            small={tiny ?? small}
            style={tiny ? styles.textWrapperIcon : styles.textWrapper}
          >
            {collectible.tokenId
              ? ` #${formatTokenId(collectible.tokenId)}`
              : ''}
          </Text>
        </View>
      ),
    [
      styles,
      style,
      cover,
      big,
      small,
      tiny,
      collectible,
      onPressColectible,
      pressNft,
    ],
  );

  const renderMedia = useCallback(() => {
    if (
      displayNftMedia ||
      (!displayNftMedia && isIpfsGatewayEnabled && isIPFSUri(sourceUri))
    ) {
      if (
        renderAnimation &&
        collectible.animation &&
        collectible.animation.includes('.mp4')
      ) {
        return (
          <MediaPlayer
            onClose={onClose}
            uri={collectible.animation}
            style={[styles.mediaPlayer, cover && styles.cover, style]}
          />
        );
      }
      if (sourceUri) {
        /*
         * the tiny boolean is used to indicate when the image is the NFT source icon
         */
        return (
          <RemoteImage
            fadeIn
            resizeMode={'contain'}
            source={{ uri: sourceUri }}
            style={[
              styles.image,
              tiny && styles.tinyImage,
              small && styles.smallImage,
              big && styles.bigImage,
              cover && styles.cover,
              style,
            ]}
            onError={fallback}
            testID="nft-image"
          />
        );
      }
    }

    if (
      !displayNftMedia ||
      (!isIpfsGatewayEnabled && !collectible.error?.startsWith('Both'))
    ) {
      return renderFallback(true);
    }

    return renderFallback(false);
  }, [
    collectible,
    sourceUri,
    onClose,
    renderAnimation,
    style,
    tiny,
    small,
    big,
    cover,
    styles,
    isIpfsGatewayEnabled,
    renderFallback,
    fallback,
    displayNftMedia,
  ]);

  return (
    <View style={styles.container(collectible.backgroundColor)}>
      {renderMedia()}
    </View>
  );
};

export default CollectibleMedia;
