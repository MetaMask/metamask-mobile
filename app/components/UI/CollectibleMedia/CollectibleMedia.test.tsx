import React from 'react';

import CollectibleMedia from './CollectibleMedia';

import renderWithProvider from '../../../util/test/renderWithProvider';
import initialBackgroundState from '../../../util/test/initial-background-state.json';

const mockInitialState = {
  engine: {
    backgroundState: initialBackgroundState,
  },
};

jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: jest.fn().mockImplementation(() => mockInitialState),
}));
const mockedNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
  };
});

describe('CollectibleMedia', () => {
  it('should render correctly', () => {
    const wrapper = renderWithProvider(
      <CollectibleMedia
        collectible={{
          name: 'NAME',
          image: 'IMAGE',
          imagePreview: 'IMAGE',
          tokenId: 123,
          address: '0x123',
          backgroundColor: 'red',
          tokenURI:
            'ipfs://QmXt7k3uoihWSyzduXErHFGTTQ3a9rnokzw9s4ywKXKsA7QmXt7k3uoihWSyzduXErHFGTTQ3a9rnokzw9s4ywKXKsA7',
        }}
      />,
      { state: mockInitialState },
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('should render collectible image if the ipfs gateway is enabled and display nft media is enabled', () => {
    const { getByTestId } = renderWithProvider(
      <CollectibleMedia
        collectible={{
          name: 'NAME',
          image: 'https://',
          imagePreview: 'https://',
          tokenId: 123,
          address: '0x123',
          backgroundColor: 'red',
          tokenURI:
            'ipfs://QmXt7k3uoihWSyzduXErHFGTTQ3a9rnokzw9s4ywKXKsA7QmXt7k3uoihWSyzduXErHFGTTQ3a9rnokzw9s4ywKXKsA7',
        }}
      />,
      { state: mockInitialState },
    );

    const fallbackCollectible = getByTestId('nft-image');
    expect(fallbackCollectible).toBeDefined();
  });
});
