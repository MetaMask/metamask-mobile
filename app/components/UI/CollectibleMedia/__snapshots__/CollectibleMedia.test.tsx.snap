// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectibleMedia should render correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#red",
      "borderRadius": 12,
      "flex": 0,
    }
  }
>
  <View
    useNativeDriver={true}
  >
    <Image
      fadeIn={true}
      onError={[Function]}
      onLoadEnd={[Function]}
      resizeMode="contain"
      source={
        {
          "uri": "IMAGE",
        }
      }
      style={
        [
          {
            "borderRadius": 12,
          },
          undefined,
          undefined,
          undefined,
          undefined,
          undefined,
        ]
      }
      testID="nft-image"
    />
    <View
      collapsable={false}
      style={
        {
          "bottom": 0,
          "left": 0,
          "opacity": 1,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
    >
      <View
        style={
          [
            {
              "borderRadius": 12,
            },
            {
              "backgroundColor": "#eee",
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;
